### DISCUSSION TITLE
I'm unable to detect when pocketbase is being Restarted by JSVM

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#3866] [Q&A]
I've had some success getting my webcam handler into the JSVM.  It will be really nice when it can be referenced with `$utils` or `$$utils`.

```javascript
  const utils = $app.cache().get("utils")
  routerAdd("GET", "/webcam", utils.webcamHandler("/dev/video0"))
```

The problem is that I want to make it a single route like this instead of it being a separate long-running goroutine process, and then having only the handler connect to that.  I mean that I want to encapsulate all of it into this easy to use single call.

The specific problem is that I'm not able to detect when the JSVM watcher has restarted the bootstrapping.  This could be thought of in terms of a very general goroutine inside of a route handler instead of the webcam aspect of it.  I'm just not seeing anywhere that I can get a context signal or an event for this restarting.

Does there need to be a new `restart` or `teardown` app event?

I've somehow managed to get as far as I have in Golang without doing any channel work that ended up being a final result.  So I'm not sure if I'm missing something basic, or if you just haven't exposed structures for this yet.

**Comment by ganigeorgiev** [comment]
I'm having difficulties understanding your use case and what you are trying to do, but there is no hook for the restart event at the moment (the auto app restart currently works only on unix with `execve`). I'll consider eventually in the future to fire the `OnTerminate` hook or create a dedicated one for the restart.

**Reply by dschissler** [reply]
The use case is that I want to be able to just define a route like this to have all goroutines handled inside of it, but the JSVM introduced a new challenge in that things are being restarted.  The route is being run again, which creates new goroutines for the camera.

I can probably handle this by recording this internally in a singleton cache store in the webcam handler package, but I think that it should also be possible to cancel things.

This was a non-issue before using JSVM.  It's just was not a thing to reload all routes on a process until I was using JSVM.  I just think that there isn't currently enough signalling to handle this behavior.

I was thinking about having this nice generic echo webcam route handle that can take a context as a signal for optional termination.  Then I'd wrap that in a thin Pocketbase code to be able to respond to an event, or a context to tell the route handler goroutine to terminate.

For now though I'll just store those goroutines inside in a cache so that they don't get started twice for the same camera.

**Reply by dschissler** [reply]
@ganigeorgiev Among other things, it seems that I was confused because the deferred call to the camera was never running upon application restart.  I was confused this whole time thinking that the application was only restarting internal structures, and not the entire application.

**Reply by dschissler** [reply]
@ganigeorgiev Ok, so after realizing that my concurrency skills suck I'm back to the beginning about the onTerminate event not being fired on restart.  I was initially expecting that event to fire, but when it didn't I started making wild assumptions about how the process wasn't actually restarting, but was just internally resetting.  I really screwed myself up thinking that.  I even looked at that damn exec code, but I hallucinated into it meaning something else, because surely it must (somehow).

So I know that you're not having much time, but I consider this to be a pretty massive bug.  People could bleed to death over it.
